rholang/src/main/scala/coop/rchain/rholang/interpreter/PrettyPrinter.scala

```diff
@@ -171,14 +171,45 @@ final case class PrettyPrinter(
 
   @SuppressWarnings(Array("org.wartremover.warts.Throw"))
   private def buildStringM(t: GeneratedMessage, indent: Int): Coeval[String] = Coeval.defer {
-    val content = t match {
+    t match {
       case v: Var => buildStringM(v)
       case s: Send =>
         buildChannelStringM(s.chan) |+| pure {
           if (s.persistent) "!!("
           else "!("
         } |+| buildSeq(s.data) |+| pure(")")
 
+      case l: Let =>
+        val (totalFree, declarationsString) = ((0, pure("")) /: l.declarations.zipWithIndex) {
+          case ((previousFree, string), (declaration, i)) =>
+            val patternString =
+              this
+                .copy(
+                  freeShift = boundShift + previousFree,
+                  boundShift = 0,
+                  freeId = boundId,
+                  baseId = setBaseId()
+                )
+                .buildParPattern(Seq(declaration.pattern))
+
+            (
+              declaration.freeCount + previousFree,
+              string |+| patternString |+| pure(" = ") |+| buildStringM(declaration.value) |+| pure {
+                if (i != l.declarations.length - 1) " , " else ""
```

Comment:
No, because we need to carry this state of the pretty printer as we build the declarations. A simple intercalation can't do that. You'd need a `foldIntercalate` or something.

Reply:
Why do you say that? I don't see any dependency between the declarations of the let-construction.

Reply:
This is _inside_ of a fold over the declaration list. The `previousFree` value is a dependency between the declarations. I need to carry that in the fold over the declarations to correctly calculate the pattern string. I may as well build the declarations string while I'm folding over them already.